### PR TITLE
feat: remove `ReactGA` and migrate to `GA4` for tracking

### LINKED ISSUE #1332: [Feature]: Support Google Analytics 4 properties
### Requirement

Starting July 1, 2023 Google will be transitioning to GA 4 properties.
The exact message on GA home page: 
```
On July 1, 2023, this property will stop processing data. Starting in March 2023, for continued website measurement, migrate your original property settings to a Google Analytics 4 (GA4) property, or they'll be copied for you to an existing GA4 property, reusing existing site tags.
```

### Problem

I created a GA4 version of my original Jaeger UI property and tried to use it but no luck. Looks like the react-ga package that is used in this repo does not support GA tags (Vs UA ones).
Found an issue in that repo: https://github.com/react-ga/react-ga/issues/541

### Proposal

I have not researched it enough, but looks like react-ga4 package might work for GA tags.
https://github.com/codler/react-ga4

### Open questions

_No response_

### PR DESCRIPTION
**Pull request by EshaanAgg**
## Which problem is this PR solving?
Fixes #1332 

## Description of the changes
- Remove the `ReactGA` library for analytics and instead load the script ourselves asynchronously
- Manually make calls to the `gtag` function with the appropriate parameters

## How was this change tested?
- NA

## Checklist
- [x] I have read https://github.com/jaegertracing/jaeger/blob/master/CONTRIBUTING_GUIDELINES.md
- [x] I have signed all commits
- [x] I have added unit tests for the new functionality
- [x] I have run lint and test steps successfully
  - for `jaeger`: `make lint test`
  - for `jaeger-ui`: `yarn lint` and `yarn test`

### CODE REVIEW

File: packages/jaeger-ui/src/utils/tracking/ga.tsx [hunk 1]
```diff
@@ -24,6 +23,40 @@ import { logTrackingCalls } from './utils';
 import { getAppEnvironment, shouldDebugGoogleAnalytics } from '../constants';
 import parseQuery from '../parseQuery';
 
+// Modify the global scope to add the `gtag` function and the `dataLayer` array
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/tracking/ga.tsx:26]: Why do we need to attach gtag function to the window object? Could we not use the same approach as in ReactGA where it had some module-level globals?

File: packages/jaeger-ui/src/utils/tracking/ga.tsx [hunk 2]
```diff
@@ -24,6 +23,40 @@ import { logTrackingCalls } from './utils';
...
+  }
+
+  const script = document.createElement('script');
+  script.async = true;
+  script.src = `${gtagUrl}?id=${GA_MEASUREMENT_ID}`;
+  document.body.appendChild(script);
+
+  window.dataLayer = window.dataLayer || [];
+  window.gtag = (...args) => {
+    window.dataLayer.push(args);
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/tracking/ga.tsx:48]: what is the "dataLayer" and why it's needed? please add comments explaining

File: packages/jaeger-ui/src/utils/tracking/ga.tsx (hunk 1 above)
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/tracking/ga.tsx:26]: based on https://developers.google.com/tag-platform/gtagjs/install, I think only the dataLayer needs to be public, the gtag function can be private in this module

File: packages/jaeger-ui/src/utils/tracking/ga.tsx [hunk 3]
```diff
@@ -24,6 +23,40 @@ import { logTrackingCalls } from './utils';
 import { getAppEnvironment, shouldDebugGoogleAnalytics } from '../constants';
 import parseQuery from '../parseQuery';
 
+// Modify the global scope to add the `gtag` function and the `dataLayer` array
+declare global {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/tracking/ga.tsx:27]: What effect does this have on other code? Will the other code start seeing `window` as having a `dataLayer` property? Why do we want that, shouldn't this knowledge be private to this module?

**Review by yurishkuro — COMMENTED** [review]
I think the approach is great! One less dependency.

File: packages/jaeger-ui/src/utils/tracking/ga.tsx (hunk 1 above)
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/ga.tsx:26]: Ahh yes, nice catch. I will update the implementation! Thanks.

File: packages/jaeger-ui/src/utils/tracking/ga.tsx (hunk 3 above)
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/ga.tsx:27]: I don't think this knowledge should remain private to this module, as the `dataLayer` attribute is globally attached to the `window` object. Keeping it private means that all the other modules can assume there is no property named `dataLayer` on `window`, and thus if needed, they can add that themselves, which will mess with our current implementation.

Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/ga.tsx:27]: On second thoughts, the most ideal implementation would be for `dataLayer` to have the appropriate types within `ga.ts` and have the type `never` outside it so that no other module tries to access or modify the same.

File: packages/jaeger-ui/src/utils/tracking/ga.test.js [hunk 4]
```diff
@@ -96,33 +97,33 @@ describe('google analytics tracking', () => {
...
     });
   });
 
   describe('trackError', () => {
     it('tracks an error', () => {
       tracking.trackError('a');
-      expect(calls).toEqual([
-        ['send', { hitType: 'exception', exDescription: expect.any(String), exFatal: false }],
+      expect(window.dataLayer).toEqual([
+        ['event', 'exception', { description: expect.any(String), fatal: false }],
```
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/tracking/ga.test.js:113]: If there is no external schema and we control the names, changing them means we're introducing breaking changes for people who may already be using the existing integration, so their dashboards could break.

File: packages/jaeger-ui/src/utils/tracking/utils.tsx [hunk 5]
```diff
@@ -12,14 +12,13 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-import ReactGA from 'react-ga';
+import { WindowWithGATracking } from './ga';
+
+declare let window: WindowWithGATracking;
 
 // eslint-disable-next-line import/prefer-default-export
 export const logTrackingCalls = () => {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/tracking/utils.tsx:20]: can we move this function inside ga.tsx (and tests)?  I don't see it being used outside, so odd to expose as public, especially since the dataLayer contact is for GA only.

File: packages/jaeger-ui/src/utils/tracking/ga.tsx [hunk 6]
```diff
@@ -24,6 +23,40 @@ import { logTrackingCalls } from './utils';
...
+export interface WindowWithGATracking extends Window {
+  dataLayer: (string | object)[][] | undefined;
+}
+
+declare let window: WindowWithGATracking;
+
+// Function to add a new event to the Google Analytics dataLayer
+const gtag = (...args: (string | object)[]) => {
+  if (window !== undefined)
+    if (window.dataLayer !== undefined) {
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/utils/tracking/ga.tsx:39]: Why are you relying on precisely `undefined` here? What about this:
```suggestion
  if (window)
    if (window.dataLayer && isArray(window.dataLayer)) {
```

File: packages/jaeger-ui/src/utils/tracking/utils.tsx (hunk 5 above)
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/tracking/utils.tsx:20]: in fact, if feels like it could be combined with the gtag() function instead of always calling it explicitly

File: packages/jaeger-ui/src/vite-env.d.ts [hunk 7]
```diff
@@ -14,3 +14,13 @@
...
+
+declare namespace global {
+  // eslint-disable-next-line @typescript-eslint/naming-convention
+  interface Window {
+    // dataLayer property is used by Google Tag Manager
+    // It should be only be changed with the help of the custom Interface WindowWithGATracking
+    // defined in utils/tracking/ga.ts
+    dataLayer: never;
+  }
+}
```
Comment (yurishkuro) [inline packages/jaeger-ui/src/vite-env.d.ts:26]: Could you elaborate why this change is needed?

File: packages/jaeger-ui/src/utils/tracking/ga.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/ga.test.js:113]: There is no definite schema that these calls adhere to. You can find the [API specification here](https://developers.google.com/tag-platform/gtagjs/reference). There are also just some `recommended events` and `recommended parameters` for these events. 

I tried to look at the source of the `ReactGA` and the API specification shared above to figure out the possible actual `tag` calls the library made for us.

File: packages/jaeger-ui/src/vite-env.d.ts (hunk 7 above)
Comment (EshaanAgg) [reply packages/jaeger-ui/src/vite-env.d.ts:26]: > On second thoughts, the most ideal implementation would be for `dataLayer` to have the appropriate types within `ga.ts` and have the type `never` outside it so that no other module tries to access or modify the same. 

This had been my initial reasoning but I think it would be better to remove it considering the way you have advised to handle the types in the previous comment.

File: packages/jaeger-ui/src/utils/tracking/utils.tsx (hunk 5 above)
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/utils.tsx:20]: Aight. The combining idea sounds nice. I'll check it out.

File: packages/jaeger-ui/src/utils/tracking/ga.tsx (hunk 6 above)
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/ga.tsx:39]: Sign catch!

File: packages/jaeger-ui/src/utils/tracking/ga.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (yurishkuro) [reply packages/jaeger-ui/src/utils/tracking/ga.test.js:112]: I an judging by the fact that the unit test is changing, so it seems like you're changing the schema. Can we keep it the same as before?

File: packages/jaeger-ui/src/utils/tracking/utils.tsx (hunk 5 above)
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/utils.tsx]: I have combined the logging with the `gtag` function itself in the latest implementation. This `WindowWithGATracking` is now only defined and used in `ga.tsx`

File: packages/jaeger-ui/src/utils/tracking/ga.test.js (hunk 4 above)
[the remark this answers is not in the document]
Comment (EshaanAgg) [reply packages/jaeger-ui/src/utils/tracking/ga.test.js:112]: Earlier, the unit tests tested what calls were being made to the `ReactGA` library. What calls, in turn, the same made to the `gtag` library was completely abstracted away from us.
Now the tests are checking what `gtag` calls are being made. Since we are losing the `ReactGA` library altogether, the tests can't be kept the same.

**Review by yurishkuro — APPROVED** [review]
Thanks!
